### PR TITLE
PN-764: Add gcp instance reaper

### PR DESCRIPTION
**Pull request by bradhe**
Title says it all. Support for GCP connections should land in Relay before we can fully integrate this workflow in production environments.

### CODE REVIEW

File: gcp-instance-reaper/filter-instances.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+import re
+
+from nebula_sdk import Interface, Dynamic as D
+
+
+ni = Interface()
+
+# The `MINUTES_TO_WAIT` global variable is the number of minutes to wait for
+# a termination_date label to appear for the GCP instance. Please note that the
+# AWS Lambdas are limited to a 5 minute maximum for their total run time.
```
Comment (kenazk) [inline gcp-instance-reaper/filter-instances.py]: > `AWS Lambdas`

Probably should just remove this sentence around lambdas.

File: gcp-instance-reaper/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+selects a subset of those to terminate. The termination criteria are:
+
+* Not tagged with a termination date or lifetime after 4 minutes
+* The `termination_date` or `lifetime` labels are present but cannot be parsed
+* The `termination_date` or `lifetime` labels indicate that the instance has
+  expired
+
+An instance may be configured to never terminate if its `lifetime` tag has the
+special value `indefinite`.
+
```
Comment (kenazk) [inline gcp-instance-reaper/README.md:15]: The most recent feedback I got from Dex (Replicated) was it wasn't clear to him that the termination criteria is actually evaluating the *value* of the tag `lifetime` or `termination_date`, not just the *existence* of the tag. Can you specify the value format expected by the tag?

File: gcp-instance-reaper/gcp-instance-reaper.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  spec:
+    terminationDateLabel: !Parameter terminationDateLabel
+    lifetimeLabel: !Parameter lifetimeLabel
+    instances: !Output {from: list-instances, name: instances}
+  inputFile: https://raw.githubusercontent.com/puppetlabs/relay-workflows/master/gcp-instance-reaper/filter-instances.py
+- name: delete-approval
+  description: Wait for approval to delete instances 
+  type: approval
+  dependsOn: filter-instances
+  when:
```
Comment (kenazk) [inline gcp-instance-reaper/gcp-instance-reaper.yaml:36]: @rick-a-lane-ii does this work now?

**Review by kenazk — COMMENTED** [review]
Added a couple comments. Otherwise looks good.

File: gcp-instance-reaper/filter-instances.py (hunk 1 above)
Comment (bradhe) [reply gcp-instance-reaper/filter-instances.py]: Good find. I thought I removed all references to AWS...

File: gcp-instance-reaper/gcp-instance-reaper.yaml (hunk 3 above)
Comment (rick-a-lane-ii) [reply gcp-instance-reaper/gcp-instance-reaper.yaml:36]: > @rick-a-lane-ii does this work now?

Approvals have been modified to support when conditions. The core of the implementation is in staging and will be deployed to production soon. With the latest changes, approvals will respect the when condition, but still wait on approval (this is unfortunate but anticipated). As expected, there are quite a few other changes that need to be addressed (approvals are currently a one-off implementation across the board, and that by itself needs addressing).

Remaining:
- Conditions should short-circuit on any known false value; it's not doing this currently (this was a likely situation, and has been officially confirmed now that I have inspected the code).
- Confirmation that the UI will respond accordingly to this situation.
- Confirmation that the one-off approval status handling in the API respects this.
- Confirmation that the status handling (in all of the backend logic) differentiates between an approval rejection and an approval being skipped due to the when conditions.

**Pull request merged**